keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+
+Credentials may also be specified at more granular levels in order to allow for context dependent keys.  E.g. When authorization differs between BucketClasses or between individual operations.  This may be facilitated by defining a set of string keys which the core automation will be aware of, so that Secrets may be referenced by BucketClasses.  For example:
+
+```yaml
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+```
+
+- **Per BucketClass:** A secret may be made specific to a BucketClass.  This suits cases where authorization may be segregated in the object store.  The Secret may then be defined explicitly in the `bucketClass.parameters` map.
+
+  ```yaml
+  cosi.io/provisioner-secret-name: "foo"
+  cosi.io/provisioner-secret-namespace: "bar"
+  ```
+
+- **Per Operation/Bucket:** Unique credentials are passed per Bucket or operation. In order to support dynamic Secret naming, templating similar to [CSI Secret templating](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html) may be used.  E.g.
+
+  ```yaml
+  "${bucket.name}"
+  "${bucket.namespace}"
+  ```
+  
+  Admins may then define a BucketClass with the following parameters included:
+
+  *Per Bucket Operation*
+  
+  ```yaml
+  cosi.io/provisioner-secret-name: "${bucket.name}"
+  cosi.io/provisioner-secret-namespace: "${bucket.namespace}"
+  ```
+---
+
+## gRPC Definitions
+
+There is one service defined by COSI - `Provisioner`. This will need to be satisfied by the vendor-provisioner in order to be COSI-compatible
+
+```
+service Provisioner {
+    rpc ProvisionerGetInfo (ProvisionerGetInfoRequest) returns (ProvisionerGetInfoResponse) {}
+
+    rpc ProvisionerCreateBucket (ProvisionerCreateBucketRequest) returns (ProvisionerCreateBucketResponse) {}
+    rpc ProvisionerDeleteBucket (ProvisionerDeleteBucketRequest) returns (ProvisionerDeleteBucketResponse) {}
+
+    rpc ProvisionerGrantBucketAccess (ProvisionerGrantBucketAccessRequest) returns (ProvisionerGrantBucketAccessResponse);
+    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
+}
+```
+
+#### ProvisionerGetInfo
+
+This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.
+
+```
+message ProvisionerGetInfoRequest {
+    // Intentionally left blank
+}
+
+message ProvisionerGetInfoResponse {    
+    string provisioner_identity = 1;
+}
+```
+
+#### ProvisonerCreateBucket
+
+This call is made to create the bucket in the backend. If the bucket already exists, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner.
+
+```
+message ProvisionerCreateBucketRequest {    
+    // This field is REQUIRED
+    string bucket_name = 1;
+
+    map<string,string> bucket_context = 4;
+
+    enum AnonymousBucketAccessMode {
+	BUCKET_PRIVATE = 0;
+	BUCKET_READ_ONLY = 1;
```

Comment:
These now no longer match the values defined at https://github.com/kubernetes/enhancements/pull/1383/files#diff-9942e2f090a0582218253969881d6409R210. Could make sense to keep them aligned, for clarity.